File: lib/engine_client/models/auth_state.dart
// Copyright (c) 2026 Osman Alperen Çinar-Koraş (oakisnotree). Licensed under AGPL-3.0.

/// Auth state received from the engine via WebSocket events.
class AuthState {
  final String mode;
  final String? tokenScope;
  final String? username;
  final String? email;

  const AuthState({
    required this.mode,
    this.tokenScope,
    this.username,
    this.email,
  });

  factory AuthState.fromJson(Map<String, dynamic> json) {
    return AuthState(
      mode: json['mode'] as String,
      tokenScope: json['token_scope'] as String?,
      username: json['username'] as String?,
      email: json['email'] as String?,
    );
  }
}

/// Auth mode constants matching the engine's enum values.
class AuthMode {
  static const undetermined = 'undetermined';
  static const anonymous = 'anonymous';
  static const connected = 'connected';
}

/// Token scope constants matching the engine's enum values.
class TokenScope {
  static const full = 'full';
  static const emailVerification = 'email_verification';
  static const partial2Fa = 'partial_2fa';
  static const setup2Fa = 'setup_2fa';
  static const awaitingBackupConfirmation = 'awaiting_backup_confirmation';
  static const deviceRegistration = 'device_registration';
}

File: packages/dspatch_engine/src/domain/models/auth_state.rs
// Copyright (c) 2026 Osman Alperen Çinar-Koraş (oakisnotree). Licensed under AGPL-3.0.

use serde::{Deserialize, Serialize};

use crate::domain::enums::{AuthMode, TokenScope};

/// The current authentication state of the app.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AuthState {
    pub mode: AuthMode,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub token: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub token_scope: Option<TokenScope>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub username: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub email: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub device_id: Option<String>,
}

impl AuthState {
    /// Local-only mode, no server connection.
    pub fn anonymous() -> Self {
        Self {
            mode: AuthMode::Anonymous,
            token: None,
            token_scope: None,
            username: None,
            email: None,
            device_id: None,
        }
    }

    /// App startup, auth not yet determined.
    pub fn undetermined() -> Self {
        Self {
            mode: AuthMode::Undetermined,
            token: None,
            token_scope: None,
            username: None,
            email: None,
            device_id: None,
        }
    }
}
